### PR TITLE
chore(#9397): prototype for documenting UI state contacts page list view access

### PR DESCRIPTION
**Pull request by lorerod**
# Description

closes https://github.com/medic/cht-core/issues/9397

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs
If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:9397-prototype-for-documenting-ui-state-contacts-page-list-view-access/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:9397-prototype-for-documenting-ui-state-contacts-page-list-view-access/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:9397-prototype-for-documenting-ui-state-contacts-page-list-view-access/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: .gitignore
Comment (latin-panda) [inline .gitignore:61]: we have other images in the repo, it can accidentally remove them. It's better to specify the folder 
```suggestion
/documentation/*.png
```

File: tests/e2e/default/contacts/list-view-login-visual.js
Comment (latin-panda) [inline tests/e2e/default/contacts/list-view-login-visual.js:1]: Is the idea to have a suite to generate documentation? 
So, it does not affect other e2e tests or break the documentation data. Then, it will keep generating the same documentation with the same data, which is good.

Comment (latin-panda) [reply tests/e2e/default/contacts/list-view-login-visual.js:1]: I wonder if this is more of a tool we execute on demand rather than every push to the CI. Then, CI will run faster while people are developing a feature. 

Or this can run for specific jobs in the CI, like when building a beta before release, then get the images, and we can compare.

In both cases, since this is not asserting, it'd fit better outside the /e2e folder 🤔

**Review by latin-panda — COMMENTED** [review]
Thanks @lorerod ! 
I'll review on Monday, but I left some thoughts and questions below

File: tests/e2e/default/contacts/list-view-login-visual.js
Comment (lorerod) [reply tests/e2e/default/contacts/list-view-login-visual.js:1]: These tests can be more of an on-demand tool than part of every CI run for now. 
How would you feel about organizing this testing code under the `/tests/e2e/visual` folder? This way, it remains separate from the core e2e tests but will remain part of them for the future, as it asserts things indirectly when running the steps.
With a separate command to run these tests only when needed, we can focus on their specific purpose, such as comparing images when building a beta before release. This will enable us to generate and compare screenshots for documentation, and eventually, we can implement visual regression testing using [WebdriverIO's visual testing service](https://webdriver.io/docs/visual-testing/).

Comment (latin-panda) [reply tests/e2e/default/contacts/list-view-login-visual.js:1]: That makes sense since we plan to compare images. Then, it also makes sense to put it in the CI because it's another test (not just generating documentation), but maybe on `push`, like this: `if: ${{ github.event_name == 'push' }}`

File: tests/factories/cht/generate.js
Comment (latin-panda) [inline tests/factories/cht/generate.js:108]: At this point, I think allPersons is a flat - regular array, so where we can do 
```suggestion
  const reports = [];
  allPersons.forEach(person => reports.add(...[
```
Just trying to make it more readable, but see if this suggestion really helps :)

File: tests/e2e/visual/contacts/list-view-login-visual.wdio-spec.js
Comment (latin-panda) [inline tests/e2e/visual/contacts/list-view-login-visual.wdio-spec.js:51]: Usually start with the "expected" then the "condition"
```suggestion
    it('should show contacts page tab when can_view_contact and can_view_contacts_tab permissions are enabled', async () => {
```

File: tests/e2e/visual/contacts/list-view-login-visual.wdio-spec.js
Comment (latin-panda) [inline tests/e2e/visual/contacts/list-view-login-visual.wdio-spec.js:53]: we usually do lowercase and dashes for filenames, can we do the same here?
```suggestion
      await generateScreenshot('contact-page', 'tab-visible');
```

**Review by latin-panda — CHANGES_REQUESTED** [review]
Great job Lore 🤩 I left some questions and suggestions below, please let me know what you think

File: tests/e2e/visual/wdio.conf.js
Comment (lorerod) [reply tests/e2e/visual/wdio.conf.js:12]: Sharp is managing the screenshot image size, not the window size.

File: tests/e2e/visual/wdio.conf.js
Comment (lorerod) [reply tests/e2e/visual/wdio.conf.js:17]: Sharp is managing the screenshot image size, not the window size.

File: tests/utils/screenshots.js
Comment (lorerod) [reply tests/utils/screenshots.js:10]: I will create another constant for this value. As I was explaining, 'setWindowSize' has a minimum limit of 500 px, so for this test, I have to use 'setViewportSize' also if the scenario is mobile.

File: tests/factories/cht/generate.js
Comment (lorerod) [reply tests/factories/cht/generate.js:82]: Full name. It's that ok?

File: tests/factories/cht/generate.js
Comment (lorerod) [reply tests/factories/cht/generate.js:87]: This is the `person doc`. We first create the `clinic` with this person's name and then create the `person` with the clinic as a parent.

File: tests/utils/screenshots.js
Comment (latin-panda) [reply tests/utils/screenshots.js]: Okay with excluding it in [the workflow](https://github.com/medic/cht-core/blob/master/.github/workflows/build.yml). IMO, as soon as we have 1 test case comparing images, it can be included in the workflow

File: tests/factories/cht/generate.js
Comment (latin-panda) [reply tests/factories/cht/generate.js:82]: Ok, I went to check [the doc site](https://docs.communityhealthtoolkit.org/building/features/contacts/), seems they use the full name

File: tests/factories/cht/generate.js
Comment (latin-panda) [reply tests/factories/cht/generate.js:87]: Ah got it. BTW I think a better name is `primaryContact`

Comment (latin-panda) [reply tests/factories/cht/generate.js:87]: Shouldn't this primaryContact have the same id as `const contact = {` in L72? 

Maybe better create the person first, then update the clinic, so for example: 

```
 const primaryContact = personFactory.build({
    .....
     ...
  });

 const clinic = placeFactory.place().build({
      type: 'clinic',
      parent: { _id: healthCenter._id, parent: healthCenter.parent },
      name: `${primaryPerson.name} Family`,
      contact: primaryContact
});

primaryContact.parent = clinic;

```

File: tests/factories/cht/generate.js
Comment (latin-panda) [reply tests/factories/cht/generate.js:97]: I'm wondering if they have the same last name since they belong to the same HH.

Comment (lorerod) [reply tests/factories/cht/generate.js:97]: I based the names in [doc site contacts page](https://docs.communityhealthtoolkit.org/building/features/contacts/)

File: tests/factories/cht/generate.js
Comment (latin-panda) [inline tests/factories/cht/generate.js:18]:
```suggestion
// Fixed collection of real-world data
const FIRST_NAMES = ['Amanda', 'Beatrice', 'Dana', 'Fatima', 'Gina', 'Helen', 'Isabelle', 'Jessica', 'Ivy', 'Sara'];
const LAST_NAMES = ['Allen', 'Bass', 'Dearborn', 'Flair', 'Gorman', 'Hamburg', 'Ivanas', 'James', 'Moore', 'Taylor'];
const PHONE_NUMBERS = [
  '+256414345783', '+256414345784', '+256414345785',
  '+256414345786', '+256414345787', '+256414345788',
  '+256414345789', '+256414345790', '+256414345791',
  '+256414345792'
];
const PATIENT_IDS = [65421, 65422, 65423, 65424, 65425, 65426, 65427, 65428, 65429, 65430];

```

File: tests/e2e/visual/contacts/list-view-login-visual.wdio-spec.js
Comment (latin-panda) [inline tests/e2e/visual/contacts/list-view-login-visual.wdio-spec.js:10]:
```suggestion
describe('Contact List Page', () => {
```

File: tests/e2e/visual/contacts/list-view-login-visual.wdio-spec.js
Comment (latin-panda) [inline tests/e2e/visual/contacts/list-view-login-visual.wdio-spec.js:71]: If this is default config, the chw role is already an offline user so no need to update it again in L14. Same with the ones below

File: tests/factories/cht/generate.js
Comment (latin-panda) [inline tests/factories/cht/generate.js:94]: When chaining functions, each function goes in a new line
```suggestion
const additionalPersons = Array
  .from({ length: nbrPersons - 1 })
  .map((_, i) => {
      const additionalPersonName = `${firstNames[i % firstNames.length]} ${lastNames[i % lastNames.length]}`;
      const additionalPhoneNumber = phoneNumbers[i % phoneNumbers.length];
      return personFactory.build({
        parent: { _id: clinic._id, parent: clinic.parent },
        name: additionalPersonName,
        patient_id: patientIds[i % patientIds.length],
        phone: additionalPhoneNumber
      });
  });

```

File: tests/factories/cht/generate.js
Comment (latin-panda) [inline tests/factories/cht/generate.js:64]: It's better to avoid big callbacks on functions and try to keep code as flat as possible. Can you try to make it a function so it's something similar to this? 
```js
const clinics = Array
  .from({ length: nbrClinics })
  .map((_, index) => createClinic(index));
```

**Review by latin-panda — CHANGES_REQUESTED** [review]
Very nice Lore! It's almost ready! I just added some minor feedback below
Thank you!

File: tests/e2e/visual/contacts/list-view-login-visual.wdio-spec.js
Comment (lorerod) [reply tests/e2e/visual/contacts/list-view-login-visual.wdio-spec.js:71]: @latin-panda I understand I need to remove L14. But I need help understanding what to do with L71. Can you please explain further?

Comment (latin-panda) [reply tests/e2e/visual/contacts/list-view-login-visual.wdio-spec.js:70]: Nothing. After removing L14, this looks good! Thanks! Resolving comment now

File: tests/e2e/visual/contacts/list-view-login-visual.wdio-spec.js
Comment (latin-panda) [inline tests/e2e/visual/contacts/list-view-login-visual.wdio-spec.js:68]: Space between test cases
```suggestion
    });
    
    it('should hide contacts page as tab and from menu option ' +
```

File: tests/factories/cht/generate.js
Comment (latin-panda) [inline tests/factories/cht/generate.js:87]: Code style: [One method per line should be used](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/#method-chaining)
```suggestion
  return Array
    .from({ length: nbrPersons - 1 })
    .map((_, i) => {
```

File: tests/factories/cht/generate.js
Comment (latin-panda) [inline tests/factories/cht/generate.js:108]:
```suggestion
  const clinicsData = Array
    .from({ length: nbrClinics })
    .map((_, index) => {
```

File: tests/factories/cht/generate.js
Comment (latin-panda) [inline tests/factories/cht/generate.js:128]: Like this no need to cache a function in a variable 
```suggestion
  if (useRealNames) { 
    return createDataWithRealNames({ healthCenter, user, nbrClinics, nbrPersons });
  }
  return createDataWithFixedData({ healthCenter, user, nbrClinics, nbrPersons });
```

File: tests/utils/screenshots.js
Comment (latin-panda) [inline tests/utils/screenshots.js:42]: minor, return early then no need of `else` block
```suggestion
  if (await isMobile()) {
    return await browser.emulateDevice({
      viewport: {
        width: MOBILE_VIEWPORT_WIDTH,
        height: MOBILE_VIEWPORT_HEIGHT,
        isMobile: true,
        hasTouch: true,
      },
      userAgent: 'Mozilla/5.0 (Linux; Android 11; Pixel 4) ' +
      'AppleWebKit/537.36 (KHTML, like Gecko) ' +
      'Chrome/92.0.4515.159 Mobile Safari/537.36'
    });
  } 
  
  return await browser.setWindowSize(DESKTOP_WINDOW_WIDTH, DESKTOP_WINDOW_HEIGHT);
```

**Review by latin-panda — APPROVED** [review]
Awesome! Thanks a lot @lorerod ✨
Just minor code style, but approving now

File: package.json
Comment (latin-panda) [inline package.json:160]: Oh, I almost forgot: once you merge, please update the tech radars to include this new technology. [We have one for contributors and one for implementers ](https://docs.communityhealthtoolkit.org/contribute/tech-radar/)

**Pull request merged**
